common/network-common/src/main/java/org/apache/spark/network/crypto/AuthEngine.java

```diff
@@ -213,7 +220,7 @@ private TransportCipher generateTransportCipher(
         transcript,  // Passing this as the HKDF salt
         OUTPUT_IV_INFO,  // This is the HKDF info field used to differentiate IV values
         AES_GCM_KEY_SIZE_BYTES);
-    SecretKeySpec sessionKey = new SecretKeySpec(sharedSecret, "AES");
+    SecretKeySpec sessionKey = new SecretKeySpec(derivedKey, "AES");
```

Comment:
Let us flag guard this change - given we want to backport to 3.x patch releases as well as giving users migration opportunities. This flag can be flipped to default in a subsequent minor version.

Reply:
I added a flag to let someone override this and revert back to the original protocol.